### PR TITLE
GHA workflow for code quality checks

### PR DESCRIPTION
**Pull request by bandish-shah**
# What does this PR do?

Adds a GitHub Actions workflow for code quality checks (pre-commit hooks).

Removes these checks from Jenkins pipelines.

# What issue(s) does this change relate to?

Resolve CO-1397.

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: .github/workflows/code-quality.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          - "3.8"
+          - "3.9"
+        pip_deps:
+          - "[all]"
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python_version }}
+      - name: Cache pip
```
Comment (dakinggg) [inline .github/workflows/code-quality.yaml:35]: How important is the cache? I'd like to catch version changes even if `setup.py` didn't change. e.g. we didn't pin a dep that we should have.

**Review by dakinggg — APPROVED** [review]
LGTM assuming all the checks pass

File: .github/workflows/code-quality.yaml (hunk 1 above)
Comment (bandish-shah) [reply .github/workflows/code-quality.yaml:35]: That's a good point, it's mostly just a runtime improvement.  I'll remove it for now and we can consider adding it back if it's considerably slower.

**Comment by dakinggg** [comment]
https://github.com/mosaicml/composer/pull/1720 to test

**Pull request merged**